lib/net.c

```diff
@@ -1547,151 +1545,109 @@ start_iface(struct gatekeeper_if *iface, unsigned int num_attempts_link_get)
 	 */
-	net_conf->gatekeeper_per_lcore_cache_size = RTE_MIN(
-		RTE_MEMPOOL_CACHE_MAX_SIZE,
-		net_conf->gatekeeper_num_mbuf / 1.5);
-	for (i = net_conf->gatekeeper_per_lcore_cache_size; i >= 1; i--) {
-		if (net_conf->gatekeeper_num_mbuf % i == 0) {
-			net_conf->gatekeeper_per_lcore_cache_size = i;
+	*cache_size = RTE_MIN(RTE_MEMPOOL_CACHE_MAX_SIZE, *num_mbuf / 1.5);
+	for (i = *cache_size; i >= 1; i--) {
+		if (*num_mbuf % i == 0) {
+			*cache_size = i;
 			break;
 		}
 	}
 
-	G_LOG(NOTICE, "net: %s: total_pkt_burst (front) = %hu packets, total_pkt_burst (back) = %hu packets, gatekeeper_num_lcores = %d lcores, gatekeeper_total_rx_desc = %hu descriptors, gatekeeper_total_tx_desc = %hu descriptors, gatekeeper_max_num_pkt = %u packets, gatekeeper_num_mbuf = %u packets, gatekeeper_per_lcore_cache_size = %u mbufs.\n",
-		__func__, net_conf->front.total_pkt_burst,
-		net_conf->back.total_pkt_burst, gatekeeper_num_lcores,
-		gatekeeper_total_rx_desc, gatekeeper_total_tx_desc,
-		gatekeeper_max_num_pkt, net_conf->gatekeeper_num_mbuf,
-		net_conf->gatekeeper_per_lcore_cache_size);
+	G_LOG(NOTICE, "%s: %s: total_pkt_burst = %hu packets, total_rx_desc = %hu descriptors, total_tx_desc = %hu descriptors, max_num_pkt = %u packets, num_mbuf = %u packets, cache_size = %u mbufs.\n",
+		block_name, __func__, total_pkt_burst, total_rx_desc,
+		total_tx_desc, max_num_pkt, *num_mbuf, *cache_size);
 }
 
-static int
-init_net_stage1(void *arg)
+struct rte_mempool *
+create_pktmbuf_pool(const char *block_name, unsigned int lcore,
+	unsigned int num_mbuf, unsigned int cache_size)
 {
-	struct net_config *net_conf = arg;
-	uint32_t i;
-
-	calculate_net_config_para(net_conf);
-
-	if (net_conf->gatekeeper_pktmbuf_pool == NULL) {
-		net_conf->gatekeeper_pktmbuf_pool =
-			rte_calloc("mbuf_pool", net_conf->numa_nodes,
-				sizeof(struct rte_mempool *), 0);
-		if (net_conf->gatekeeper_pktmbuf_pool == NULL) {
-			G_LOG(ERR, "net: %s: out of memory\n", __func__);
-			return -1;
-		}
-	}
-
-	/* Initialize pktmbuf pool on each used NUMA node. */
-	for (i = 0; i < net_conf->numa_nodes; i++) {
-		char pool_name[64];
-		int ret;
-
-		if (!net_conf->numa_used[i] ||
-				net_conf->gatekeeper_pktmbuf_pool[i] != NULL)
-			continue;
-
-		ret = snprintf(pool_name, sizeof(pool_name), "pktmbuf_pool_%u",
-			i);
-		RTE_VERIFY(ret > 0 && ret < (int)sizeof(pool_name));
-		net_conf->gatekeeper_pktmbuf_pool[i] =
-			rte_pktmbuf_pool_create(pool_name,
-				net_conf->gatekeeper_num_mbuf,
-				net_conf->gatekeeper_per_lcore_cache_size,
-				sizeof(struct list_head),
-				RTE_MBUF_DEFAULT_BUF_SIZE, (unsigned)i);
+	struct rte_mempool *mp;
+	char pool_name[64];
+	int ret = snprintf(pool_name, sizeof(pool_name), "pktmbuf_pool_%s_%u",
+		block_name, lcore);
+	RTE_VERIFY(ret > 0 && ret < (int)sizeof(pool_name));
+	mp = rte_pktmbuf_pool_create_by_ops(pool_name, num_mbuf, cache_size,
+		sizeof(struct list_head), RTE_MBUF_DEFAULT_BUF_SIZE,
+		rte_lcore_to_socket_id(lcore), "ring_mp_sc");
```

Comment:
Now that we use `"ring_mp_sc"`, shouldn't `cache_size` always be zero?

Reply:
If so, please remember to remove the corresponding parameter from the wiki once this pull request is merged.

Reply:
According to [rte_mempool_create()](https://doc.dpdk.org/api/rte__mempool_8h.html#a503f2f889043a48ca9995878846db2fd), the purpose of maintaining a per-lcore object cache is to limit the accesses to the common lockless pool. However, in our case, we only have a single consumer. I agree that we should set ``cache_size`` to be zero. I will update our wiki once this pull request is merged.